src/detectors/BTOF/BTOFHitDigi.cc

```diff
@@ -0,0 +1,196 @@
+// SPDX-License-Identifier: LGPL-3.0-or-later
+// Copyright (C) 2024 Souvik Paul, Chun Yuen Tsang, Prithwish Tribedy
+// Special Acknowledgement: Kolja Kauder
+
+// A general digitization for BToFHit from simulation
+// 1. Smear energy deposit with a/sqrt(E/GeV) + b + c/E or a/sqrt(E/GeV) (relative value)
+// 2. Digitize the energy with dynamic ADC range and add pedestal (mean +- sigma)
+// 3. Time conversion with smearing resolution (absolute value)
+// 4. Signal is summed if the SumFields are provided
+//
+// Author: Souvik Paul, Chun Yuen Tsang
+// Date: 19/07/2024
+
+#include "DD4hep/Detector.h"
+#include "DDRec/Surface.h"
+#include "TF1.h"
+#include "TMath.h"
+#include <Evaluator/DD4hepUnits.h>
+#include <TGraph.h>
+#include <bitset>
+#include <fmt/format.h>
+#include <iostream>
+#include <vector>
+
+#include "BTOFHitDigi.h"
+#include "Math/SpecFunc.h"
+#include "algorithms/digi/BTOFHitDigiConfig.h"
+
+// using namespace dd4hep;
+// using namespace dd4hep::Geometry;
+// using namespace dd4hep::DDRec;
+// using namespace eicrecon;
+using namespace dd4hep::xml;
+
+namespace eicrecon {
+
+std::vector<bool> BTOFHitDigi::ToDigitalCode(int value, int numBits) {
+  std::bitset<32> binary(value); // Convert integer to binary representation
+  std::vector<bool> digitalCode;
+
+  for (int i = numBits - 1; i >= 0; --i) {
+    digitalCode.push_back(binary.test(i));
+  }
+
+  return digitalCode;
+}
+
+void BTOFHitDigi::init(const dd4hep::Detector* detector, std::shared_ptr<spdlog::logger>& logger) {
+  // m_detector = detector;
+  m_log = logger;
+
+  adc_range = pow(2, adc_bit);
+  tdc_range = pow(2, tdc_bit);
+
+  // using juggler internal units (GeV, mm, radian, ns)
+  tRes    = m_cfg.tRes / dd4hep::ns;
+  stepTDC = dd4hep::ns / m_cfg.resolutionTDC;
+
+  _neighborFinder.init(detector);
+}
+
+double BTOFHitDigi::_integralGaus(double mean, double sd, double low_lim, double up_lim) {
+  double up  = -0.5 * ROOT::Math::erf(TMath::Sqrt(2) * (mean - up_lim) / sd);
+  double low = -0.5 * ROOT::Math::erf(TMath::Sqrt(2) * (mean - low_lim) / sd);
+  return up - low;
+}
+
+std::unique_ptr<edm4eic::RawTrackerHitCollection>
+BTOFHitDigi::execute(const edm4hep::SimTrackerHitCollection* simhits) {
+  // auto rawhits = std::make_unique<edm4eic::RawTrackerHitCollection>();
+  auto rawhits = std::make_unique<edm4eic::RawTrackerHitCollection>();
+  // const auto [sim_hits] = simhits;
+
+  // find the hits that belong to the same group (for merging)
+  std::unordered_map<uint64_t, std::vector<std::size_t>> merge_map;
```

Comment:
You can have a map to the hit pointers themselves instead of storing and recalling the hits by vertex

Reply:
That part of my code is refactored. It's now located in src/detector/BTOF/TOFPulseGeneration.cc